frontends/common/constantFolding.cpp

```diff
@@ -1108,15 +1108,44 @@ const IR::Node *DoConstantFolding::postorder(IR::SelectExpression *expression) {
     return result;
 }
 
+class FilterLikelyAnnot : public Transform {
+    IR::Statement *preorder(IR::Statement *stmt) {
+        prune();
+        return stmt;
+    }
+    IR::BlockStatement *preorder(IR::BlockStatement *stmt) {
+        prune();
+        visit(stmt->annotations, "annotations");
+        return stmt;
+    }
+    IR::Annotation *preorder(IR::Annotation *annot) {
+        prune();
+        if (annot->name == IR::Annotation::likelyAnnotation) return nullptr;
+        if (annot->name == IR::Annotation::unlikelyAnnotation) {
+            warning(ErrorType::WARN_IGNORE, "ignoring %1% on always taken statement", annot);
+            return nullptr;
+        }
+        return annot;
+    }
+};
+
 const IR::Node *DoConstantFolding::postorder(IR::IfStatement *ifstmt) {
     if (auto cond = ifstmt->condition->to<IR::BoolLiteral>()) {
         if (cond->value) {
-            return ifstmt->ifTrue;
+            if (auto blk = ifstmt->ifFalse ? ifstmt->ifFalse->to<IR::BlockStatement>() : nullptr) {
+                if (auto annot = blk->getAnnotation(IR::Annotation::likelyAnnotation))
+                    warning(ErrorType::WARN_IGNORE, "ignoring %1% on never taken statement", annot);
```

Comment:
Won't it be too verbose? I can easily see that unlikely case could appear after some e.g. generic routine would be inlined and then condition will be constfolded on e.g. on error path? Something like this:
```
function generic() {
  if (not error)  @likely {
   dosomething();
  } else {
    dootherthing();
  }
}

function handle_error() {
  generic();
}
```

As a comparison, C/C++ compilers do not produce diagnostics while constfolding `__builtin_expect` and friends.

Reply:
Trying to find a happy medium between too verbose and not verbose enough.  @vlstill requested this warning, and it is just a warning.  Perhaps leave it and see if anyone complains?